hail/src/main/scala/is/hail/expr/types/virtual/TNDArray.scala

```diff
@@ -45,6 +45,38 @@ final case class TNDArray(elementType: Type, nDimsBase: NatBase, override val re
     sb.append("]")
   }
 
+  override def str(a: Annotation): String = {
+    if (a == null) "NA" else {
+      val a_row = a.asInstanceOf[Row]
+      val shape = a_row(2).asInstanceOf[Row].toSeq.asInstanceOf[Seq[Long]].map(_.toInt)
+      val data = a_row(4).asInstanceOf[UnsafeIndexedSeq]
+
+      def dataToNestedString(data: Iterator[Annotation], shape: Seq[Int], sb: StringBuilder):Unit  = {
+        if (shape.isEmpty) {
+          sb.append(data.next().toString)
+        }
+        else {
+          sb.append("[")
+          val howMany = shape.head
+          (0 until howMany).foreach{repeat =>
```

Comment:
not a big deal, but this is probably more efficiently written as a c-style for loop, to avoid materializing `0 until howMany`